src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -419,11 +399,27 @@ generateP scrutTy (pat, alt)
         require (annotOfExp alt) $ CEquals datS t
 
         (alt',sub) <- generateX alt
+        let altTy'  = annResult (annotOfExp alt')
+        altTy      <- TypeVar <$> fresh
+        altTp      <- TypeVar <$> fresh
+
+        -- Require alternative types to have the same temporality if they
+        -- do have temporalities. Otherwise defaults to TemporalityPure.
+        require (annotOfExp alt) $ CExtractTemporality altTp altTy altTy'
 
-        e' <- lift $ lift State.get
+        -- Require the alternative types without temporality to be the same.
+        require (annotOfExp alt) $ CEquals resTy altTy
```

Comment:
ah, I think "requireData" would suffice here, which checks just the data type as opposed to the temporality and possibility too:
https://github.com/ambiata/icicle/blob/case/src/Icicle/Source/Checker/Constraint.hs#L287